### PR TITLE
port Registry I/O

### PR DESCRIPTION
**Pull request by ShaMan123**
## Description

ports what remains from #8282 to #8500

## Changes

- renamed: `setClass` to `setJSONClass`, `getClass` to `getJSONClass`
- exposed `setClass`: register class to both JSON and SVG registeries
- changed signature of `getXXXClass` accepting a `strict` param indicating if it should throw or not
- use registry to resolve `enlivenObjectEnlivenables`
- changed `getJSONClass` signature to accept an object, needed to resolve Gradient in `enlivenObjectEnlivenables`
- refactored `enlivenObjectEnlivenables` to return an object **only** with enlived stuff, previously was returning everything
- changed Gradient to fit the registry
-  enliven Shadow in `enlivenObjectEnlivables` not in `set` (I don't think it is breaking but devs can add `type: shadow` to data in case they fall into an edge case)
- fix `fromObject` regression 3fd06607d

## Gist

closes #8282

### CODE REVIEW

**Review by ShaMan123 — COMMENTED** [review]
### Changes
- renamed: `setClass` to `setJSONClass`, `getClass` to `getJSONClass`
- exposed `setClass`: register class to both JSON and SVG registeries
- changed signature of `getXXXClass` accepting a `strict` param indicating if it should throw or not
- use registry to resolve `enlivenObjectEnlivenables`
- change `getJSONClass` signature to accept an object, needed to resolve Gradient in `enlivenObjectEnlivenables`
- fixed canvas class when serializing

@asturur take a look
These are what changes I wanted to port from #8282 
Besides that there are semantics in the registry architecture that are negligent.
I think we should consider the registry be part of the future context object passed around in fabric so the dev could impl a different registry entirely and register it to the context, as long as they stick to the API. That can wait.

**Comment by ShaMan123** [comment]
BTW I think the registry should sit directly under `src`

**Review by ShaMan123 — COMMENTED** [review]
updated from master and rebased

**Review by ShaMan123 — COMMENTED** [review]
updated from master

**Review by ShaMan123 — COMMENTED** [review]
updated from master

File: src/gradient/gradient.class.ts [hunk 1]
```diff
@@ -368,11 +376,14 @@ export class Gradient<
...
     el: SVGGradientElement,
-    instance: FabricObject,
-    svgOptions: SVGOptions
-  ): Gradient<GradientType> {
+    callback: (instance: Gradient<'linear' | 'radial'>) => any,
+    {
+      object: { left, top },
+      ...svgOptions
+    }: SVGOptions & { object: FabricObject }
+  ) {
```
Comment (ShaMan123) [inline src/gradient/gradient.class.ts:384]: made the signature identical to the rest of fabric

File: src/shapes/Object/Object.ts [hunk 2]
```diff
@@ -1850,9 +1849,10 @@ export class FabricObject<
       ...options
     }: { extraParam?: string; signal?: AbortSignal } = {}
   ): Promise<FabricObject> {
-    return enlivenObjectEnlivables<any>(cloneDeep(object), options).then(
+    const data = cloneDeep(object);
+    return enlivenObjectEnlivables<any>(data as any, options).then(
       (enlivedMap) => {
-        const allOptions = { ...options, ...enlivedMap };
+        const allOptions = { ...data, ...enlivedMap };
```
Comment (ShaMan123) [inline src/shapes/Object/Object.ts:1855]: an important fix

File: src/util/misc/objectEnlive.ts [hunk 3]
```diff
@@ -128,52 +113,39 @@ export const enlivenObjects = (
...
-            return pattern;
-          });
-      }
-      return value;
+    const promises = Object.values(serializedObject).map(async (value) => {
+      const klass =
+        value &&
+        typeof value === 'object' &&
+        classRegistry.getJSONClass(value, false);
+      if (!klass) return;
```
Comment (ShaMan123) [inline src/util/misc/objectEnlive.ts:137]: love this

File: test/unit/gradient.js [hunk 4]
```diff
@@ -74,6 +74,12 @@
     });
   }
 
+  function fromElement(element, object, options = {}) {
+    let gradient;
+    fabric.Gradient.fromElement(element, g => (gradient = g), { ...options, object });
+    return gradient;
+  }
+
```
Comment (ShaMan123) [inline test/unit/gradient.js:82]: lazy test adaption

**Review by ShaMan123 — COMMENTED** [review]
no features here

File: src/shapes/Image.ts [hunk 5]
```diff
@@ -718,21 +718,19 @@ export class Image extends FabricObject {
...
+      filters ? enlivenObjects<BaseFilter>(filters, options) : undefined,
       enlivenObjectEnlivables(object, options),
-    ]).then(([el, filters = [], [resizeFilter] = [], hydratedProps = {}]) => {
+    ]).then(([el, filters = [], hydratedProps = {}]) => {
       return new this(el, {
         ...object,
         src,
         crossOrigin,
         filters,
-        resizeFilter,
```
Comment (asturur) [inline src/shapes/Image.ts:735]: i already reverted this once, i don't understand where resizeFilters gets enlived now.

Comment (ShaMan123) [reply src/shapes/Image.ts:735]: in enlivenObjectEnlivenables, it is a prop in the map that can be handled by the method as opposed to filters that is an array of stuff to enliven under a prop in the map.

**Review by ShaMan123 — COMMENTED** [review]
updated from master

ready 

maybe I should expose `getClass` alias to `getJSONClass`

File: test/visual/z_svg_export.js [hunk 6]
```diff
@@ -386,7 +386,7 @@
...
...or":"rgb(0,52,95)","opacity":1},{"offset":0,"color":"rgb(0,68,115)","opacity":1}],"offsetX":-51.096,"offsetY":-38.088,"gradientTransform":[1,0,0,-1,-50,111]},"points":[{"x":199,"y":211.912},{"x":199,"y":211.912},{"x":199,"y":174.746},{"x":84.498,"y":117.723},{"x":84.498,"y":58.96},{"x":51.096,"y":38.088},{"x":51.096,"y":138.597}]},{"type":"polygon","version":"3.1.0","left":-74.5,"top":-123.52,"width":147.84,"height":56.93,"fill":{"type":"linear","coords":{"x1":174.922,"y1":110.6136,"x2":174.922,"y2":-135.0903},"colorStops":[{"offset":1,"color":"rgb(0,38,57)","opacity":1},{"offset":0,"color":"rgb(0,46,59)","opacity":1}],"offsetX":-51,"offsetY":-1.985,"gradientTransform":[1,0,0,-1,-50,111]},"stroke":{"type":"linear","coords":{"x1":174.922,"y1":110.6136,"x2":174.922,"y2":-135.0903},"colorStops":[{"offset":1,"color":"rgb(0,38,57)","opacity":1},{"offset":0,"color":"rgb(0,46,59)","opacity":1}],"offsetX":-51,"offsetY":-1.985,"gradientTransform":[1,0,0,-1,-50,111]},"points":[{"x":84.396,"y":58.904},{"x":84.396,"y":58.892},{"x":124.939,"y":39.118},{"x":165.485,"y":58.892},{"x":198.844,"y":38.046},{"x":124.912,"y":1.985},{"x":51,"y":38.035},{"x":51,"y":38.067},{"x":84.368,"y":58.918}]}]}]}';
```
Comment (ShaMan123) [inline test/visual/z_svg_export.js:389]: added `type: shadow` to shadow json

**Review by ShaMan123 — COMMENTED** [review]
fixed enliven shadow
previously was done in `_set`

**Review by ShaMan123 — COMMENTED** [review]
updated

**Review by ShaMan123 — COMMENTED** [review]
updated, resolved conflicts successfully I hope

**Comment by ShaMan123** [comment]
@asturur what about this??

**Review by ShaMan123 — COMMENTED** [review]
updated from master and resolved conflicts

**Pull request closed without merging**